### PR TITLE
Feature/mesh volume

### PR DESCRIPTION
**Pull request by RasmusHogslatt**
Changes proposed in this PR:
 * Support abuffer rendering with meshes and volumes simultaneously.

### CODE REVIEW

**Comment by martinfalk** [comment]
Refactored the mesh and volume rasterization and merged it into the OIT module. 
* separate processor `MeshVolumeRenderer` handling both volume and other rasterization objects
* supports concurrent rendering of max 4 volume rasterizations plus other rasterizations (mesh, line, spheres). Limit can be increased if necessary.

---
image

File: modules/oit/glsl/oit/abufferlinkedlist.glsl [hunk 1]
```diff
@@ -72,12 +72,34 @@ layout(std430, binding = 7) buffer abufferStorage {
...
+MESH:
+data.x: uint prev;
+data.y: float depth;
+data.z: float alpha;
+data.w: float RGB (10bits each) + 1 bit unused + 1 bit type
+
+VOLUME:
+data.x: uint prev;
+data.y: float depth;
+data.z: R(20bits) + G(12bits)
```
Comment (petersteneteg) [inline modules/oit/glsl/oit/abufferlinkedlist.glsl]: RGB here is really position.x position.y, position.z right?

File: modules/oit/glsl/oit/volumeresolve.frag [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+//     ...
+//     case n:
+//         scalar = sampleVolume(volumeSamplers[n], volumeIndex, samplePosWorld);
+//         break;
+//     default:
+//         break;
+// }
+// ```
+
+uniform sampler3D volumeSamplers[4];
```
Comment (petersteneteg) [inline modules/oit/glsl/oit/volumeresolve.frag]: could have added a define for "4"

File: modules/oit/glsl/oit/volumeresolve.frag [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+                                scalar = sampleVolume(volumeSamplers[2], volumeIndex, samplePosWorld);
+                                break;
+                            case 3:
+                                scalar = sampleVolume(volumeSamplers[3], volumeIndex, samplePosWorld);
+                                break;
+                            default:
+                                break;
+                        }
+
+                        vec4 color = applyTF(tfSamplers[volumeIndex], scalar);
```
Comment (petersteneteg) [inline modules/oit/glsl/oit/volumeresolve.frag:272]: why is tfSamplers[volumeIndex], any different from volumeSamplers[0]? both should require constants expressions?

File: modules/oit/glsl/volumerasterizer.frag [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+#endif
+
+uniform int volumeId;
+
+void main() {
+    vec4 fragColor = vec4(1.0, 0.0, 0.0, 1.0);
+
+#ifdef USE_FRAGMENT_LIST
+    // fragment list rendering
+        if (volumeId >= 0) {
```
Comment (petersteneteg) [inline modules/oit/glsl/volumerasterizer.frag]: indentation

File: modules/oit/include/modules/oit/datastructures/rasterization.h [hunk 5]
```diff
@@ -80,6 +84,14 @@ class IVW_MODULE_OIT_API Rasterization {
      */
     std::optional<mat4> boundingBox() const;
 
+    struct RaycastingState {
+        std::shared_ptr<TFLookupTable> tfLookup;
+        int channel;
+        float opacityScaling;
```
Comment (petersteneteg) [inline modules/oit/include/modules/oit/datastructures/rasterization.h]:
```suggestion
        int channel = 0;
        float opacityScaling = 0.0f;

```

File: modules/oit/include/modules/oit/datastructures/rasterization.h [hunk 6]
```diff
@@ -58,6 +61,7 @@ class IVW_MODULE_OIT_API Rasterization {
     Rasterization(Rasterization&&) = delete;
     Rasterization& operator=(const Rasterization&) = delete;
     Rasterization& operator=(Rasterization&&) = delete;
+    virtual ~Rasterization() = default;
```
Comment (petersteneteg) [inline modules/oit/include/modules/oit/datastructures/rasterization.h]: Why are you making this class polymorphic?

File: modules/oit/include/modules/oit/processors/rasterizer.h [hunk 7]
```diff
@@ -81,6 +89,7 @@ class IVW_MODULE_OIT_API Rasterizer : public Processor {
 
 private:
     RasterizeHandle handle_;
+    std::shared_ptr<Rasterization> rasterization_;
```
Comment (petersteneteg) [inline modules/oit/include/modules/oit/processors/rasterizer.h:91]: why?

File: modules/oit/include/modules/oit/datastructures/rasterization.h [hunk 8]
```diff
@@ -80,6 +84,14 @@ class IVW_MODULE_OIT_API Rasterization {
      */
     std::optional<mat4> boundingBox() const;
 
+    struct RaycastingState {
```
Comment (petersteneteg) [inline modules/oit/include/modules/oit/datastructures/rasterization.h]: should this not be in the Rasterizer class instead? 

why do we need to have it here?

File: modules/oit/include/modules/oit/datastructures/rasterization.h [hunk 9]
```diff
@@ -80,6 +84,14 @@ class IVW_MODULE_OIT_API Rasterization {
      */
     std::optional<mat4> boundingBox() const;
 
+    struct RaycastingState {
+        std::shared_ptr<TFLookupTable> tfLookup;
+        int channel;
+        float opacityScaling;
+        std::shared_ptr<const Volume> volume;
+    };
+    virtual std::optional<RaycastingState> getRaycastingState() const;
```
Comment (petersteneteg) [inline modules/oit/include/modules/oit/datastructures/rasterization.h]: remove virtual

File: modules/oit/src/processors/volumerasterizer.cpp [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        auto transform = CompositeTransform(boundingMesh->getModelMatrix(),
+                                            boundingMesh->getWorldMatrix() * worldMatrixTransform);
+        utilgl::setShaderUniforms(shader_, transform, "geometry");
+
+        MeshDrawerGL::DrawObject drawer{boundingMesh->getRepresentation<MeshGL>(),
+                                        boundingMesh->getDefaultMeshInfo()};
+        drawer.draw();
+    }
+
+    tfLookup_ = std::make_shared<TFLookupTable>(tf_.tf_.get());
```
Comment (petersteneteg) [inline modules/oit/src/processors/volumerasterizer.cpp]: Why do you need a new one each time?

**Review by petersteneteg — CHANGES_REQUESTED** [review]

File: modules/oit/glsl/oit/volumeresolve.frag (hunk 3 above)
Comment (martinfalk) [reply modules/oit/glsl/oit/volumeresolve.frag:272]: Good catch. The switch-case was a left-over from passing sampler indices to the function. Now samplers are forwarded.

File: modules/oit/glsl/oit/volumeresolve.frag (hunk 2 above)
Comment (martinfalk) [reply modules/oit/glsl/oit/volumeresolve.frag]: added `MAX_SUPPORTED_VOLUMES` which is set by the VolumeFragmentListRenderer.

File: modules/oit/glsl/oit/abufferlinkedlist.glsl (hunk 1 above)
Comment (martinfalk) [reply modules/oit/glsl/oit/abufferlinkedlist.glsl]: yes, fixed comment.

File: modules/oit/include/modules/oit/processors/rasterizer.h (hunk 7 above)
Comment (martinfalk) [reply modules/oit/include/modules/oit/processors/rasterizer.h:91]: For setting the volume ID uniforms the renderer needs to match each rasterizer with the correct ID.

https://github.com/inviwo/inviwo/blob/5afed4df451df4a678e6448a811cb148c22c56b6/modules/oit/src/processors/rasterizer.cpp#L59-L61

https://github.com/inviwo/inviwo/blob/5afed4df451df4a678e6448a811cb148c22c56b6/modules/oit/src/processors/meshvolumerenderer.cpp#L223-L233

File: modules/oit/include/modules/oit/datastructures/rasterization.h (hunk 8 above)
Comment (martinfalk) [reply modules/oit/include/modules/oit/datastructures/rasterization.h]: Moved to separate header (`raycastingstate.h`)

File: modules/oit/glsl/oit/volumeresolve.frag [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+            float tStart = ceil(depthToWorld(camera, screenPos, depth) / tIncr) * tIncr;
+            // tStart = depthToWorld(camera, screenPos, depth);
+            float tEnd = depthToWorld(camera, screenPos, nextFragDepth);
+
+            // integrate ray segments of all volumes between current depth and next fragment in world space
+            float t = tStart;
+            while (t < tEnd) {
+                const vec3 samplePosWorld = camera.position + rayDir * t;
+
+                for(int volumeIndex = 0; volumeIndex < 2; ++volumeIndex) {
```
Comment (petersteneteg) [inline modules/oit/glsl/oit/volumeresolve.frag]: 2 here should be a define?

File: modules/oit/include/modules/oit/processors/rasterizer.h (hunk 7 above)
Comment (petersteneteg) [reply modules/oit/include/modules/oit/processors/rasterizer.h:91]: ahh ic. Would it not be easier to use the Rasterizer processor for id instead?

**Review by petersteneteg — DISMISSED** [review]

File: modules/oit/include/modules/oit/processors/rasterizer.h (hunk 7 above)
Comment (martinfalk) [reply modules/oit/include/modules/oit/processors/rasterizer.h:91]: made changes to use the identifier instead.

Comment (petersteneteg) [reply modules/oit/include/modules/oit/processors/rasterizer.h:91]: Remove the ptr?

**Review by petersteneteg — APPROVED** [review]

**Pull request merged**